### DISCUSSION TITLE
Geometry Shader

### DISCUSSION THREAD

**Discussion by salexx** [StereoKit/StereoKit#1114] [Q&A] [answered]
Hello. Is it possible to use Geometry Shader in StereoKit? I haven't found any examples.

I want to visualize normals like in the image:

![normal](https://github.com/user-attachments/assets/69079fd2-549f-43bf-98c4-789a25a13e1b)

**Comment by maluoi** [comment]
SK does not support geometry shaders! However, this is generally pretty easy using SK's line renderer.

```csharp
static void DrawNormals(Vertex[] verts, float normScale)
{
    foreach(Vertex v in verts)
    {
        Lines.Add(v.position, v.position+v.normal*normScale, Color.White, normScale*0.05f);
    }
}

// Then you can call it with your mesh data. Note that GetVerts is slow, and you should only call it once!

Vertex[] verts = mesh.GetVerts();
DrawNormals(verts, 0.01f);
```

Untested code, but should get you pretty far :)

**Comment by salexx** [comment]
Thanks for the quick reply.

DrawNormals(verts, 0.01f); - works!

Too bad geometry shaders are not supported.

My geometry shaders code compiled without errors. But it seems it is not executed from StereoKit.

Maybe there is some way to execute my shader. I understand that lighting and materials from StereoKit will not work. And it will be necessary to pass the world and view matrices to this shader.

Thanks.

**Reply by maluoi** [reply, marked as answer]
Geometry shaders are generally considered to be poor performers, and are typically not used by modern pipelines. StereoKit does not have any code that can hook up a geometry shader to the rendering pipeline, so a successfully compiled geometry shader will unfortunately not get you any further!

You might be able to use something like ComputeSharp if the result does not need rendered immediately? Though ComputeSharp is more for compute shaders.

**Reply by salexx** [reply]
Thanks for the quick reply.

ComputeSharp - looks fascinating.

Will need to study it carefully )

For me, the most unclear thing in the shader is this:

o.view_id = id % sk_view_count;
id = id / sk_view_count;

where can I learn more about this code?
I think this is related to the specifics of XR rendering. For stereo output, or am I mistaken?

**Reply by maluoi** [reply]
StereoKit does a lot of instanced rendering, and yeah, this is critical for the stereo part of rendering! For single pass rendering StereoKit submits two instances, one for the left eye and one for the right. This code separates the instance index into the correct instance data array index, and the view_id for the current eye's display surface!

So if you draw a cube 10 times, StereoKit will have 20 instances. Odd instances will be left eye, even instances will be right. There will only be data for each of the 10 draws that were originally submitted, so you can find the correct instance data by dividing the instance index by 2 :)

This is done using variables in case there's more or less than 2 "eyes", like the simulator (1 view) or something like the Varjo (4 views).

I don't have strong shader docs yet, largely because I suspect that shaders could undergo some changes in the future. Right now the best shader reference is from SKs built in shaders in this repo!
